File: public/proxy.py
"""
Flask Web Proxy Server
Forwards requests to target URLs, useful for bypassing CORS restrictions.
"""

from flask import Flask, request, Response
import requests

app = Flask(__name__)

# Configure allowed headers to forward
EXCLUDED_HEADERS = [
    'content-encoding',
    'content-length',
    'transfer-encoding',
    'connection',
    'host'
]


@app.route('/proxy', methods=['GET', 'POST', 'PUT', 'DELETE', 'PATCH'])
def proxy():
    """
    Proxy endpoint that forwards requests to the target URL.

    Usage:
        GET /proxy?url=https://api.example.com/data
        POST /proxy?url=https://api.example.com/data (with JSON body)
    """
    target_url = request.args.get('url')

    if not target_url:
        return {'error': 'Missing "url" query parameter'}, 400

    # Forward headers (excluding problematic ones)
    headers = {
        key: value for key, value in request.headers
        if key.lower() not in EXCLUDED_HEADERS
    }

    try:
        # Forward the request to the target URL
        resp = requests.request(
            method=request.method,
            url=target_url,
            headers=headers,
            data=request.get_data(),
            params={k: v for k, v in request.args.items() if k != 'url'},
            cookies=request.cookies,
            allow_redirects=False,
            timeout=30
        )

        # Build response headers (excluding problematic ones)
        response_headers = [
            (name, value) for name, value in resp.raw.headers.items()
            if name.lower() not in EXCLUDED_HEADERS
        ]

        # Create response with CORS headers
        response = Response(resp.content, resp.status_code, response_headers)
        response.headers['Access-Control-Allow-Origin'] = '*'
        response.headers['Access-Control-Allow-Methods'] = 'GET, POST, PUT, DELETE, PATCH, OPTIONS'
        response.headers['Access-Control-Allow-Headers'] = '*'

        return response

    except requests.exceptions.Timeout:
        return {'error': 'Request timed out'}, 504
    except requests.exceptions.RequestException as e:
        return {'error': f'Proxy error: {str(e)}'}, 502


@app.route('/proxy', methods=['OPTIONS'])
def proxy_options():
    """Handle CORS preflight requests."""
    response = Response()
    response.headers['Access-Control-Allow-Origin'] = '*'
    response.headers['Access-Control-Allow-Methods'] = 'GET, POST, PUT, DELETE, PATCH, OPTIONS'
    response.headers['Access-Control-Allow-Headers'] = '*'
    return response


@app.route('/health')
def health():
    """Health check endpoint."""
    return {'status': 'ok'}


if __name__ == '__main__':
    app.run(host='0.0.0.0', port=5000, debug=True)

File: server/api/proxy.ts
/**
 * Web Proxy API Route
 * Forwards requests to target URLs, useful for bypassing CORS restrictions.
 *
 * Usage:
 *   GET /api/proxy?url=https://api.example.com/data
 *   POST /api/proxy?url=https://api.example.com/data (with JSON body)
 */

export default defineEventHandler(async (event) => {
  const query = getQuery(event)
  const targetUrl = query.url as string

  if (!targetUrl) {
    throw createError({
      statusCode: 400,
      message: 'Missing "url" query parameter'
    })
  }

  const method = event.method

  // Handle CORS preflight
  if (method === 'OPTIONS') {
    setResponseHeaders(event, {
      'Access-Control-Allow-Origin': '*',
      'Access-Control-Allow-Methods': 'GET, POST, PUT, DELETE, PATCH, OPTIONS',
      'Access-Control-Allow-Headers': '*'
    })
    return null
  }

  try {
    // Get request body for non-GET requests
    let body: any = undefined
    if (method !== 'GET' && method !== 'HEAD') {
      body = await readBody(event)
    }

    // Forward headers (excluding problematic ones)
    const excludedHeaders = ['host', 'connection', 'content-length']
    const headers: Record<string, string> = {}
    const requestHeaders = getHeaders(event)

    for (const [key, value] of Object.entries(requestHeaders)) {
      if (!excludedHeaders.includes(key.toLowerCase()) && value) {
        headers[key] = value as string
      }
    }

    // Build URL with additional query params (excluding 'url')
    const url = new URL(targetUrl)
    for (const [key, value] of Object.entries(query)) {
      if (key !== 'url' && value) {
        url.searchParams.set(key, value as string)
      }
    }

    // Make the request
    const response = await $fetch.raw(url.toString(), {
      method: method as any,
      headers,
      body: body ? JSON.stringify(body) : undefined,
      redirect: 'follow'
    })

    // Set CORS headers on response
    setResponseHeaders(event, {
      'Access-Control-Allow-Origin': '*',
      'Access-Control-Allow-Methods': 'GET, POST, PUT, DELETE, PATCH, OPTIONS',
      'Access-Control-Allow-Headers': '*'
    })

    return response._data

  } catch (error: any) {
    console.error('Proxy error:', error)

    throw createError({
      statusCode: error.statusCode || 502,
      message: `Proxy error: ${error.message || 'Unknown error'}`
    })
  }
})
